File: donater-bot/handlers/review.js
import { Composer } from 'grammy';
import { reviewCreate, orderById, ordersRecentForUser } from '../database/db.js';
import { notifyAdminsReview } from '../utils/notify.js';

export const reviewComposer = new Composer();

reviewComposer.command('otziv', async (ctx) => {
  try {
    const uid = ctx.from?.id;
    if (!uid) return;

    const recent = ordersRecentForUser(uid, 1);
    ctx.session.reviewOrderId = recent[0]?.id || null;

    ctx.session.step = 'await_review_text';
    await ctx.reply(ctx.i18n('write_review'), {
      parse_mode: 'HTML',
      reply_markup: {
        inline_keyboard: [
          [{ text: ctx.i18n('cancel'), callback_data: 'rev:cancel' }],
        ],
      },
    });
  } catch (e) {
    console.error(e);
    await ctx.reply(ctx.i18n('error_generic'));
  }
});

reviewComposer.callbackQuery('rev:cancel', async (ctx) => {
  try {
    await ctx.answerCallbackQuery();
    ctx.session.step = null;
    await ctx.reply(ctx.i18n('review_cancel'));
  } catch (e) {
    console.error(e);
  }
});

reviewComposer.on('message:text', async (ctx, next) => {
  try {
    if (ctx.session.step !== 'await_review_text') return next();
    const text = (ctx.message.text || '').trim();
    if (!text || text.startsWith('/')) return next();

    const uid = ctx.from.id;
    const rid = reviewCreate({
      orderId: ctx.session.reviewOrderId,
      userTgId: uid,
      text,
    });

    ctx.session.step = null;
    await ctx.reply(ctx.i18n('review_sent'), { parse_mode: 'HTML' });

    const order = ctx.session.reviewOrderId
      ? orderById(ctx.session.reviewOrderId)
      : null;
    const fakeOrder =
      order ||
      ({
        id: '—',
        username: ctx.from.username,
        product_label: '—',
        game_name: '—',
      });
    await notifyAdminsReview(
      ctx.api,
      { id: rid, text },
      fakeOrder,
      ctx.i18n
    );
  } catch (e) {
    console.error(e);
    await ctx.reply(ctx.i18n('error_generic'));
  }
});

File: donater-bot/handlers/review.py
import logging

from aiogram import F, Router
from aiogram.filters import Command
from aiogram.fsm.context import FSMContext
from aiogram.types import CallbackQuery, InlineKeyboardButton, Message
from aiogram.utils.keyboard import InlineKeyboardBuilder

from config import config
from database import queries
from states.order import ReviewFlow
from utils import tg_api
from utils.formatter import escape_html

log = logging.getLogger(__name__)
router = Router(name="review")


@router.message(Command("otziv"))
async def cmd_otziv(message: Message, t, state: FSMContext):
    uid = message.from_user.id
    if not await queries.user_has_accepted_order(uid):
        await message.answer(t("review_no_orders"))
        return
    recent = await queries.orders_recent_for_user(uid, 1)
    oid = recent[0]["id"] if recent else None
    await state.update_data(review_order_id=oid)
    await state.set_state(ReviewFlow.waiting_text)
    b = InlineKeyboardBuilder()
    b.row(InlineKeyboardButton(text=t("confirm_no"), callback_data="rev:cancel"))
    await message.answer(t("write_review"), parse_mode="HTML", reply_markup=b.as_markup())


@router.callback_query(F.data == "rev:cancel")
async def cb_rev_cancel(query: CallbackQuery, t, state: FSMContext):
    await query.answer()
    await state.clear()
    await query.message.answer(t("review_cancel"))


@router.message(ReviewFlow.waiting_text, F.text)
async def msg_review_text(message: Message, t, state: FSMContext):
    text = (message.text or "").strip()
    if text.startswith("/"):
        return
    uid = message.from_user.id
    data = await state.get_data()
    oid = data.get("review_order_id")
    rid = await queries.review_create(oid, uid, text)
    await state.clear()
    await message.answer(t("review_sent"), parse_mode="HTML")

    order = await queries.order_by_id(oid) if oid else None
    fake = order or {
        "id": "—",
        "username": message.from_user.username,
        "product_label": "—",
        "game_name": "—",
    }
    card = t(
        "review_admin_card",
        username=escape_html(fake.get("username") or "—"),
        order_id=fake.get("id", "—"),
        product=escape_html(fake.get("product_label", "—")),
        game=escape_html(fake.get("game_name", "—")),
        text=escape_html(text[:500]),
    )
    kb = tg_api.review_admin_keyboard_raw(rid, t)
    for aid in config.admin_ids:
        try:
            await tg_api.raw_send_message(aid, card, reply_markup=kb, parse_mode="HTML")
        except Exception as exc:
            log.warning("review notify %s: %s", aid, exc)
